narwhals/_ibis/utils.py

```diff
@@ -0,0 +1,221 @@
+            return dtypes.List(native_to_narwhals_dtype(ibis_dtype.value_type, version))
+    if is_struct(ibis_dtype):
+        return dtypes.Struct(
+            [
+                dtypes.Field(name, native_to_narwhals_dtype(dtype, version))
+                for name, dtype in ibis_dtype.items()
+            ]
+        )
+    if ibis_dtype.is_decimal():  # pragma: no cover
+        return dtypes.Decimal()
+    if ibis_dtype.is_time():
+        return dtypes.Time()
+    if ibis_dtype.is_binary():
+        return dtypes.Binary()
+    return dtypes.Unknown()  # pragma: no cover
+
+
+def is_interval(obj: IbisDataType) -> TypeIs[ibis_dtypes.Interval]:
+    return obj.is_interval()
+
+
+def is_array(obj: IbisDataType) -> TypeIs[ibis_dtypes.Array[Any]]:
+    return obj.is_array()
+
+
+def is_struct(obj: IbisDataType) -> TypeIs[ibis_dtypes.Struct]:
+    return obj.is_struct()
+
+
+def narwhals_to_native_dtype(  # noqa: C901, PLR0912
+    dtype: DType | type[DType], version: Version
+) -> IbisDataType:
+    dtypes = version.dtypes
+
+    if isinstance_or_issubclass(dtype, dtypes.Decimal):  # pragma: no cover
+        return ibis_dtypes.Decimal()
+    if isinstance_or_issubclass(dtype, dtypes.Float64):
+        return ibis_dtypes.Float64()
+    if isinstance_or_issubclass(dtype, dtypes.Float32):
+        return ibis_dtypes.Float32()
+    if isinstance_or_issubclass(dtype, dtypes.Int128):  # pragma: no cover
+        msg = "Int128 not supported by Ibis"
+        raise NotImplementedError(msg)
+    if isinstance_or_issubclass(dtype, dtypes.Int64):
+        return ibis_dtypes.Int64()
+    if isinstance_or_issubclass(dtype, dtypes.Int32):
+        return ibis_dtypes.Int32()
+    if isinstance_or_issubclass(dtype, dtypes.Int16):
+        return ibis_dtypes.Int16()
+    if isinstance_or_issubclass(dtype, dtypes.Int8):
+        return ibis_dtypes.Int8()
+    if isinstance_or_issubclass(dtype, dtypes.UInt128):  # pragma: no cover
+        msg = "UInt128 not supported by Ibis"
+        raise NotImplementedError(msg)
+    if isinstance_or_issubclass(dtype, dtypes.UInt64):
+        return ibis_dtypes.UInt64()
+    if isinstance_or_issubclass(dtype, dtypes.UInt32):
+        return ibis_dtypes.UInt32()
+    if isinstance_or_issubclass(dtype, dtypes.UInt16):
+        return ibis_dtypes.UInt16()
+    if isinstance_or_issubclass(dtype, dtypes.UInt8):
+        return ibis_dtypes.UInt8()
+    if isinstance_or_issubclass(dtype, dtypes.String):
+        return ibis_dtypes.String()
+    if isinstance_or_issubclass(dtype, dtypes.Boolean):
+        return ibis_dtypes.Boolean()
+    if isinstance_or_issubclass(dtype, dtypes.Categorical):
+        msg = "Categorical not supported by Ibis"
+        raise NotImplementedError(msg)
+    if isinstance_or_issubclass(dtype, dtypes.Datetime):
+        return ibis_dtypes.Timestamp()
+    if isinstance_or_issubclass(dtype, dtypes.Duration):
+        return ibis_dtypes.Interval(unit=dtype.time_unit)  # type: ignore[arg-type,unused-ignore]
```

Comment:
Thanks for following up @rwhitten577 (https://github.com/narwhals-dev/narwhals/pull/2000#discussion_r2072385686)

If you run into any of these, use the `pyright` ignore instead

```suggestion
        return ibis_dtypes.Interval(unit=dtype.time_unit)  # pyright: ignore[reportArgumentType]
```

I suppose this might be more common with `ibis` since `mypy` doesn't seem to know what's happening 😄

Reply:
Nice, thanks for the tip! Updated